src/notevennexus/validators.py

```diff
@@ -0,0 +1,269 @@
+
+
+class units_invalid(Validator):
+    def __init__(self) -> None:
+        super().__init__("units_invalid", "Invalid units attribute")
+
+    def applies_to(self, node: Dataset | Group) -> bool:
+        return isinstance(node, Dataset) and 'units' in node.attrs
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        units = node.attrs['units']
+        if units.startswith('NX_'):  # Placeholder from NeXus standard
+            return Violation(node.name, f"Invalid units {units}")
+        invalid = ['hz']
+        if units in invalid:
+            return Violation(node.name, f"Invalid units {units}")
+
+
+class index_has_units(Validator):
+    def __init__(self) -> None:
+        super().__init__(
+            "index_has_units", "Index or mask should not have units attribute"
+        )
+
+    def applies_to(self, node: Dataset | Group) -> bool:
+        names = [
+            'cue_index',
+            'cylinders',
+            'detector_faces',
+            'detector_number',
+            'event_id',
+            'event_index',
+            'faces',
+            'image_key',
+            'winding_order',
+        ]
+        name = node.name.split('/')[-1]
+        return isinstance(node, Dataset) and name in names
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        if 'units' in node.attrs:
+            return Violation(node.name)
+
+
+class mask_has_units(Validator):
+    def __init__(self) -> None:
+        super().__init__("mask_has_units", "Mask should not have units attribute")
+
+    def applies_to(self, node: Dataset | Group) -> bool:
+        return isinstance(node, Dataset) and node.name.startswith('pixel_mask')
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        if 'units' in node.attrs:
+            return Violation(node.name)
+
+
+class float_dataset_units_missing(Validator):
+    def __init__(self) -> None:
+        super().__init__(
+            "float_dataset_units_missing", "Float dataset should have units attribute"
+        )
+
+    def applies_to(self, node: Dataset | Group) -> bool:
+        return isinstance(node, Dataset) and node.dtype in [np.float32, np.float64]
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        if 'units' not in node.attrs:
+            return Violation(node.name)
+
+
+class non_numeric_dataset_has_units(Validator):
+    def __init__(self) -> None:
+        super().__init__(
+            "non_numeric_dataset_has_units",
+            "Non-numeric dataset should not have units attribute",
+        )
+
+    def applies_to(self, node: Dataset | Group) -> bool:
+        return isinstance(node, Dataset) and not np.issubdtype(node.dtype, np.number)
+
+    def validate(self, node: Dataset | Group) -> Violation | None:
+        if 'units' in node.attrs:
+            return Violation(node.name)
+
+
+def is_transformation(node: Dataset | Group) -> bool:
+    return 'transformation_type' in node.attrs
```

Comment:
Could you ever have a file where `'transformation_type'` is msising from a field that was meant to be a transformation?

Reply:
Of course you can have such a file :)
We could add another check, ensuring that all targets of any `depends_on` are proper transformations.